mqtt/mqtt-bridge/src/bridge.rs

```diff
@@ -1,31 +1,473 @@
+            self.system_address, client_id
+        );
+
+        self.connect(
+            &self.forwards,
+            self.system_address.as_str(),
+            *self.connection_settings.keep_alive(),
+            self.connection_settings.clean_session(),
+            &Credentials::Anonymous(client_id),
+        )
+        .await
+    }
+
+    async fn connect(
+        &self,
+        topics: &HashMap<String, Topic>,
+        address: &str,
+        keep_alive: Duration,
+        clean_session: bool,
+        credentials: &Credentials,
+    ) -> Result<(), BridgeError> {
+        let mut topic_filters = Vec::new();
+        for val in topics.values() {
+            topic_filters.push(TopicMapper {
+                topic_settings: val.clone(),
+                topic_filter: TopicFilter::from_str(val.pattern())
+                    .map_err(BridgeError::TopicFilterParseError)?,
+            });
+        }
+
+        let mut client = MqttClient::new(
+            address,
+            keep_alive,
+            clean_session,
+            MessageHandler::new(topic_filters, BATCH_SIZE),
+            credentials,
+        );
+
+        let subscriptions: Vec<String> = topics.keys().map(|key| key.into()).collect();
+        debug!("subscribe to remote {:?}", subscriptions);
+
+        client
+            .subscribe(subscriptions)
+            .await
+            .map_err(BridgeError::SubscribeError)?;
+
+        //TODO: handle this with shutdown
+        let _events_task = tokio::spawn(client.handle_events());
+
+        Ok(())
+    }
+}
+
+#[derive(Clone)]
+struct TopicMapper {
+    topic_settings: Topic,
+    topic_filter: TopicFilter,
+}
+
+/// Handle events from client and saves them with the forward topic
+#[derive(Clone)]
+struct MessageHandler<'a, T>
+where
+    T: Persist<'a>,
+{
+    topic_mappers: Vec<TopicMapper>,
+    inner: T,
+    phantom: PhantomData<&'a T>,
+}
+
+impl<'a, T> MessageHandler<'a, T>
+where
+    T: Persist<'a>,
+{
+    pub fn new(topic_mappers: Vec<TopicMapper>, batch_size: usize) -> Self {
+        Self {
+            topic_mappers,
+            inner: T::new(batch_size),
+            phantom: PhantomData,
+        }
+    }
+
+    fn transform(&self, topic_name: &str) -> Option<String> {
+        for mapper in &self.topic_mappers {
+            let forward_topic = mapper
+                .topic_settings
+                .local()
+                // maps if local does not have a value it uses the topic that was received,
+                // else it checks that the received topic starts with local prefix and removes the local prefix
+                .map_or(Some(topic_name.to_owned()), |local_prefix| {
+                    let prefix = format!("{}/", local_prefix);
+                    if topic_name.to_owned().starts_with(&prefix) {
+                        let rs: String = topic_name
+                            .strip_prefix(&prefix)
+                            .unwrap_or(&topic_name)
```

Comment:
I think you don't need to check starts_with to remove prefix. strip_prefix will do the same for you